{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# default_exp softwares.fastq.kallisto"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# kallisto"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "from pybiotools4p.softwares.base import Base, modify_cmd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "class Kallisto(Base):\n",
    "    def __init__(self, software, fd):\n",
    "        super(Kallisto, self).__init__(software)\n",
    "        self._default = fd\n",
    "\n",
    "    def cmd_version(self):\n",
    "        '''\n",
    "        :return:\n",
    "        '''\n",
    "        return 'echo {repr} ;echo $({software} version)'.format(\n",
    "            repr=self.__repr__(),\n",
    "            software=self._software\n",
    "        )\n",
    "\n",
    "\n",
    "    #TODO:\n",
    "    @modify_cmd\n",
    "    def cmd_inspect():\n",
    "        '''\n",
    "Inspects and gives information about an index\n",
    "        \n",
    "        '''\n",
    "        pass\n",
    "    \n",
    "    \n",
    "    #TODO:\n",
    "    @modify_cmd\n",
    "    def cmd_h5dump():\n",
    "        '''\n",
    "Converts HDF5-formatted results to plaintext\n",
    "        '''\n",
    "        pass\n",
    "    \n",
    "    \n",
    "    \n",
    "    #TODO:\n",
    "    @modify_cmd\n",
    "    def cmd_merge():\n",
    "        '''\n",
    "Merges several batch runs\n",
    "        '''\n",
    "        pass\n",
    "    \n",
    "    #TODO:\n",
    "    @modify_cmd\n",
    "    def cmd_pseudo():\n",
    "        '''\n",
    "Runs the pseudoalignment step\n",
    "        '''\n",
    "        pass\n",
    "    \n",
    "    \n",
    "    #TODO:\n",
    "    @modify_cmd\n",
    "    def cmd_bus():\n",
    "        '''\n",
    "Generate BUS files for single-cell data\n",
    "        '''\n",
    "        pass\n",
    "    \n",
    "    \n",
    "    \n",
    "    @modify_cmd\n",
    "    def cmd_quant(self, fastq,index,outdir,gtf):\n",
    "        '''\n",
    "Runs the quantification algorithm\n",
    "        '''\n",
    "        return r'''\n",
    "{software} {quant_para} \\\n",
    "        --gtf {gtf} \\\n",
    "        --output-dir {outdir} \\\n",
    "        --index {index} \\\n",
    "        {fastq}\n",
    "        '''.format(\n",
    "            software=self._software,\n",
    "            quant_para=self._default['quant'],\n",
    "            **locals()\n",
    "        )\n",
    "    \n",
    "    @modify_cmd\n",
    "    def cmd_index(self, fasta, index):\n",
    "        '''\n",
    "Builds a kallisto index\n",
    "        '''\n",
    "        return r'''\n",
    "{software} {index_para} \\\n",
    "        --index {index} \\\n",
    "        {fasta}\n",
    "        '''.format(\n",
    "            software=self._software,\n",
    "            index_para=self._default['index'],\n",
    "            **locals()\n",
    "        )\n",
    "        \n",
    "    \n",
    "    \n",
    "\n",
    "    def __repr__(self):\n",
    "        return 'kallisto:' + self._software\n",
    "\n",
    "    def __str__(self):\n",
    "        return '''\n",
    "kallisto is a program for quantifying abundances of transcripts from bulk and single-cell RNA-Seq data,\n",
    "or more generally of target sequences using high-throughput sequencing reads.\n",
    "\n",
    "When using this program in your research, please cite\n",
    "\n",
    "  Bray, N. L., Pimentel, H., Melsted, P. & Pachter, L.\n",
    "  Near-optimal probabilistic RNA-seq quantification,\n",
    "  Nature Biotechnology 34, 525-527(2016), doi:10.1038/nbt.3519\n",
    "        '''"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
