---

title: featurecounts


keywords: fastai
sidebar: home_sidebar



nb_path: "07_featurecounts.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 07_featurecounts.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="Featurecounts" class="doc_header"><code>class</code> <code>Featurecounts</code><a href="https://github.com/btrspg/pybiotools4p/tree/master/pybiotools4p/softwares/bam/featurecounts.py#L11" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>Featurecounts</code>(<strong><code>software</code></strong>, <strong><code>fd</code></strong>) :: <a href="/pybiotools4p/base.html#Base"><code>Base</code></a></p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">configparser</span>

<span class="n">config</span><span class="o">=</span><span class="n">configparser</span><span class="o">.</span><span class="n">ConfigParser</span><span class="p">()</span>
<span class="n">config</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;pybiotools4p/default.ini&#39;</span><span class="p">)</span>



<span class="n">featurecounts</span><span class="o">=</span><span class="n">Featurecounts</span><span class="p">(</span><span class="s1">&#39;featureCounts&#39;</span><span class="p">,</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;featurecounts&#39;</span><span class="p">])</span>
<span class="n">featurecounts</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>featureCounts:featureCounts</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">reference</span><span class="o">=</span><span class="s1">&#39;biology-test-data/fasta/Homo_sapiens.GRCh38.dna.primary_assembly.chromosome22.fa&#39;</span>
<span class="n">gtf_file</span><span class="o">=</span><span class="s1">&#39;biology-test-data/gtf/HS.22.gtf&#39;</span>

<span class="n">outbam_pe</span><span class="o">=</span><span class="s1">&#39;./pybiotools/hisat_pe.bam&#39;</span>
<span class="n">outbam_se</span><span class="o">=</span><span class="s1">&#39;./pybiotools/hisat_se.bam&#39;</span>
<span class="n">bams</span> <span class="o">=</span> <span class="p">[</span><span class="n">outbam_pe</span><span class="p">,</span><span class="n">outbam_se</span><span class="p">]</span>
<span class="n">outgtf</span><span class="o">=</span><span class="s1">&#39;./pybiotools/stringtie.gtf&#39;</span>
<span class="n">abundance</span><span class="o">=</span><span class="s1">&#39;./pybiotools/stringtie.txt&#39;</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">featurecounts</span><span class="o">.</span><span class="n">cmd_version</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>&#39;echo featureCounts:featureCounts ;featureCounts -v&#39;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">featurecounts</span><span class="o">.</span><span class="n">cmd_featurecounts</span><span class="p">(</span><span class="n">bams</span><span class="p">,</span> <span class="n">gtf_file</span><span class="p">,</span><span class="n">outgtf</span><span class="p">,</span> <span class="s1">&#39;./tmp&#39;</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>featureCounts -T 12 -p -t exon -g gene_id -M --ignoreDup --extraAttributes gene_name \
        -a biology-test-data/gtf/HS.22.gtf \
        -o ./pybiotools/stringtie.gtf \
        --tmpDir ./tmp \
        ./pybiotools/hisat_pe.bam ./pybiotools/hisat_se.bam
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

