---

title: hisat2

keywords: fastai
sidebar: home_sidebar



---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 04_hisat2.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="Hisat2" class="doc_header"><code>class</code> <code>Hisat2</code><a href="https://github.com/btrspg/pybiotools/tree/master/pybiotools/softwares/fastq/hisat2.py#L12" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>Hisat2</code>(<strong><code>software</code></strong>, <strong><code>fd</code></strong>) :: <a href="/pybiotools/base#Base"><code>Base</code></a></p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">configparser</span>

<span class="n">config</span><span class="o">=</span><span class="n">configparser</span><span class="o">.</span><span class="n">ConfigParser</span><span class="p">()</span>
<span class="n">config</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;pybiotools/default.ini&#39;</span><span class="p">)</span>

<span class="kn">from</span> <span class="nn">pybiotools.softwares.bam.samtools</span> <span class="kn">import</span> <span class="n">Samtools</span>

<span class="n">samtools</span><span class="o">=</span><span class="n">Samtools</span><span class="p">(</span><span class="s1">&#39;samtools&#39;</span><span class="p">,</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;samtools&#39;</span><span class="p">])</span>
<span class="n">hisat2</span><span class="o">=</span><span class="n">Hisat2</span><span class="p">(</span><span class="s1">&#39;hisat2&#39;</span><span class="p">,</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;hisat2&#39;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">reference</span><span class="o">=</span><span class="s1">&#39;biology-test-data/fasta/Homo_sapiens.GRCh38.dna.primary_assembly.chromosome22.fa&#39;</span>
<span class="n">gtf_file</span><span class="o">=</span><span class="s1">&#39;biology-test-data/gtf/HS.22.gtf&#39;</span>
<span class="n">genome_index_prefix</span><span class="o">=</span><span class="s1">&#39;./pybiotools/hisat2_index&#39;</span>

<span class="n">fq1</span><span class="o">=</span><span class="s1">&#39;./biology-test-data/fastq/HS.22.r1.fq.gz&#39;</span>
<span class="n">fq2</span><span class="o">=</span><span class="s1">&#39;./biology-test-data/fastq/HS.22.r2.fq.gz&#39;</span>
<span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;./pybiotools/HS_22&#39;</span>
<span class="n">snp_file</span><span class="o">=</span><span class="s1">&#39;/path/to/snp&#39;</span>
<span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;./pybiotools/tmp&#39;</span>
<span class="n">genome_ss</span><span class="o">=</span><span class="s1">&#39;./pybiotools/tmp.ss&#39;</span>
<span class="n">genome_exon</span><span class="o">=</span><span class="s1">&#39;./pybiotools/tmp.exon&#39;</span>
<span class="n">summary_pe</span><span class="o">=</span><span class="s1">&#39;./pybiotools/pe.summary.txt&#39;</span>
<span class="n">summary_se</span><span class="o">=</span><span class="s1">&#39;./pybiotools/se.summary.txt&#39;</span>
<span class="n">samtools_idx</span><span class="o">=</span><span class="s1">&#39;biology-test-data/fasta/Homo_sapiens.GRCh38.dna.primary_assembly.chromosome22.fa.fai&#39;</span>
<span class="n">outbam_pe</span><span class="o">=</span><span class="s1">&#39;./pybiotools/hisat_pe.bam&#39;</span>
<span class="n">outbam_se</span><span class="o">=</span><span class="s1">&#39;./pybiotools/hisat_se.bam&#39;</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">hisat2</span><span class="o">.</span><span class="n">cmd_version</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>&#39;echo hisat2:hisat2 ;hisat2 --version&#39;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">hisat2</span><span class="o">.</span><span class="n">cmd_build_index</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="n">genome_index_prefix</span><span class="p">,</span> <span class="n">genome_ss</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">genome_exon</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                        <span class="n">genome_genotype</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">genome_snp</span><span class="o">=</span><span class="kc">None</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>hisat2-build -p 16 biology-test-data/fasta/Homo_sapiens.GRCh38.dna.primary_assembly.chromosome22.fa  ./pybiotools/hisat2_index
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">hisat2</span><span class="o">.</span><span class="n">cmd_prepare_snp_ucsc</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span><span class="n">snp_file</span><span class="p">,</span><span class="n">prefix</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>awk &#39;BEGIN{OFS=&#34;\t&#34;} {if($2 ~ /^chr/) {$2 = substr($2, 4)}; if($2 == &#34;M&#34;) {$2 = &#34;MT&#34;} print}&#39; /path/to/snp \
    &gt; ./pybiotools/tmp_snp.tmp
hisat2_extract_snps_haplotypes_UCSC.py biology-test-data/fasta/Homo_sapiens.GRCh38.dna.primary_assembly.chromosome22.fa ./pybiotools/tmp_snp.tmp ./pybiotools/tmp
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">hisat2</span><span class="o">.</span><span class="n">cmd_prepare_exon_ss</span><span class="p">(</span><span class="n">gtf_file</span><span class="p">,</span><span class="n">prefix</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>hisat2_extract_splice_sites.py biology-test-data/gtf/HS.22.gtf &gt; ./pybiotools/tmp.ss
hisat2_extract_exons.py biology-test-data/gtf/HS.22.gtf &gt; ./pybiotools/tmp.exon
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">hisat2</span><span class="o">.</span><span class="n">cmd_build_index</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="n">genome_index_prefix</span><span class="o">+</span><span class="s1">&#39;_ss_exon&#39;</span><span class="p">,</span> <span class="n">genome_ss</span><span class="o">=</span><span class="n">genome_ss</span><span class="p">,</span> <span class="n">genome_exon</span><span class="o">=</span><span class="n">genome_exon</span><span class="p">,</span>
                        <span class="n">genome_genotype</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">genome_snp</span><span class="o">=</span><span class="kc">None</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>hisat2-build -p 16 biology-test-data/fasta/Homo_sapiens.GRCh38.dna.primary_assembly.chromosome22.fa  --ss ./pybiotools/tmp.ss --exon ./pybiotools/tmp.exon ./pybiotools/hisat2_index_ss_exon
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">hisat2</span><span class="o">.</span><span class="n">cmd_align</span><span class="p">(</span> <span class="n">genome_index_prefix</span><span class="p">,</span> <span class="n">fq1</span><span class="p">,</span> <span class="n">fq2</span><span class="p">,</span> <span class="n">summary_pe</span><span class="p">,</span> <span class="n">samtools</span><span class="p">,</span> <span class="n">samtools_idx</span><span class="p">,</span> <span class="n">outbam_pe</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>
hisat2 --dta -5 10 -3 10 -p 16 -x ./pybiotools/hisat2_index -1 ./biology-test-data/fastq/HS.22.r1.fq.gz -2 ./biology-test-data/fastq/HS.22.r2.fq.gz --summary-file ./pybiotools/pe.summary.txt | samtools view -bSt biology-test-data/fasta/Homo_sapiens.GRCh38.dna.primary_assembly.chromosome22.fa.fai - | samtools sort - -o ./pybiotools/hisat_pe.bam
samtools index ./pybiotools/hisat_pe.bam
 
            
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">hisat2</span><span class="o">.</span><span class="n">cmd_align</span><span class="p">(</span> <span class="n">genome_index_prefix</span><span class="p">,</span> <span class="n">fq1</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">summary_se</span><span class="p">,</span> <span class="n">samtools</span><span class="p">,</span> <span class="n">samtools_idx</span><span class="p">,</span> <span class="n">outbam_se</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>
hisat2 --dta -5 10 -3 10 -p 16 -x ./pybiotools/hisat2_index -U ./biology-test-data/fastq/HS.22.r1.fq.gz  --summary-file ./pybiotools/se.summary.txt | samtools view -bSt biology-test-data/fasta/Homo_sapiens.GRCh38.dna.primary_assembly.chromosome22.fa.fai - | samtools sort - -o ./pybiotools/hisat_se.bam
samtools index ./pybiotools/hisat_se.bam
 
            
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

