---

title: STAR

keywords: fastai
sidebar: home_sidebar



---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 03_star.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="Star" class="doc_header"><code>class</code> <code>Star</code><a href="https://github.com/btrspg/pybiotools4p/tree/master/pybiotools4p/softwares/fastq/star.py#L12" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>Star</code>(<strong><code>software</code></strong>, <strong><code>fd</code></strong>) :: <a href="/pybiotools4p/base#Base"><code>Base</code></a></p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">configparser</span>

<span class="n">config</span><span class="o">=</span><span class="n">configparser</span><span class="o">.</span><span class="n">ConfigParser</span><span class="p">()</span>
<span class="n">config</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;pybiotools4p/default.ini&#39;</span><span class="p">)</span>

<span class="n">star</span><span class="o">=</span><span class="n">Star</span><span class="p">(</span><span class="s1">&#39;STAR&#39;</span><span class="p">,</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;star&#39;</span><span class="p">])</span>
<span class="n">star</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>star:STAR</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">star</span><span class="o">.</span><span class="n">cmd_version</span><span class="p">())</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>echo star:STAR;STAR --version
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">star</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Spliced Transcripts Alignment to a Reference
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">reference</span><span class="o">=</span><span class="s1">&#39;biology-test-data/fasta/Homo_sapiens.GRCh38.dna.primary_assembly.chromosome22.fa&#39;</span>
<span class="n">gtf</span><span class="o">=</span><span class="s1">&#39;biology-test-data/gtf/HS.22.gtf&#39;</span>
<span class="n">star_index_dir</span><span class="o">=</span><span class="s1">&#39;./pybiotools/star_index&#39;</span>
<span class="n">read_length</span><span class="o">=</span><span class="mi">99</span>
<span class="n">fq1</span><span class="o">=</span><span class="s1">&#39;./biology-test-data/fastq/HS.22.r1.fq.gz&#39;</span>
<span class="n">fq2</span><span class="o">=</span><span class="s1">&#39;./biology-test-data/fastq/HS.22.r2.fq.gz&#39;</span>
<span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;./pybiotools/HS_22&#39;</span>

<span class="nb">print</span><span class="p">(</span><span class="n">star</span><span class="o">.</span><span class="n">cmd_build_index</span><span class="p">(</span><span class="n">star_index_dir</span><span class="p">,</span> <span class="n">reference</span><span class="p">,</span> <span class="n">gtf</span><span class="p">,</span> <span class="n">read_length</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>STAR --runThreadN 16 --runMode genomeGenerate \
        --genomeDir ./pybiotools/star_index \
        --genomeFastaFiles biology-test-data/fasta/Homo_sapiens.GRCh38.dna.primary_assembly.chromosome22.fa \
        --sjdbGTFfile biology-test-data/gtf/HS.22.gtf \
        --sjdbOverhang 99
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">star</span><span class="o">.</span><span class="n">cmd_align</span><span class="p">(</span><span class="n">star_index_dir</span><span class="p">,</span> <span class="n">fq1</span><span class="p">,</span> <span class="n">fq2</span><span class="p">,</span> <span class="n">prefix</span><span class="p">,</span> <span class="n">gtf</span><span class="p">,</span> <span class="n">read_length</span><span class="p">,</span><span class="n">miRNA</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>STAR --runThreadN 16 \
--outSAMstrandField intronMotif \
--readFilesCommand zcat \
--outSAMtype BAM SortedByCoordinate \
    --genomeDir ./pybiotools/star_index \
    --readFilesIn ./biology-test-data/fastq/HS.22.r1.fq.gz ./biology-test-data/fastq/HS.22.r2.fq.gz \
    --outFileNamePrefix ./pybiotools/HS_22 \
    --sjdbGTFfile biology-test-data/gtf/HS.22.gtf  \
    --sjdbOverhang 99
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">star</span><span class="o">.</span><span class="n">cmd_align</span><span class="p">(</span><span class="n">star_index_dir</span><span class="p">,</span> <span class="n">fq1</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">prefix</span><span class="o">+</span><span class="s1">&#39;_mirna&#39;</span><span class="p">,</span> <span class="n">gtf</span><span class="p">,</span> <span class="n">read_length</span><span class="p">,</span><span class="n">miRNA</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>STAR --runThreadN 16 \
--outSAMstrandField intronMotif \
--readFilesCommand zcat \
--outSAMtype BAM SortedByCoordinate \
    --genomeDir ./pybiotools/star_index \
    --readFilesIn ./biology-test-data/fastq/HS.22.r1.fq.gz  \
    --outFileNamePrefix ./pybiotools/HS_22_mirna \
    --sjdbGTFfile biology-test-data/gtf/HS.22.gtf  \
    --sjdbOverhang 99 --alignEndsType EndToEnd \
--outFilterMismatchNmax 1 \
--outFilterMultimapScoreRange 0 \
--quantMode TranscriptomeSAM GeneCounts \
--outReadsUnmapped Fastx  \
--outFilterMultimapNmax 10 \
--outSAMunmapped Within \
--outFilterScoreMinOverLread 0 \
--outFilterMatchNminOverLread 0 \
--outFilterMatchNmin 16 \
--alignSJDBoverhangMin 1000 \
--alignIntronMax 1 \
--outWigType wiggle \
--outWigStrand Stranded \
--outWigNorm RPM
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

